R/dataframe-s3-base.R

```diff
@@ -121,3 +121,245 @@ head.polars_data_frame <- function(x, n = 6L, ...) x$head(n = n)
 
 #' @exportS3Method utils::tail
 tail.polars_data_frame <- function(x, n = 6L, ...) x$tail(n = n)
+
+#' Try to match `tibble` behavior as much as possible, following
+#' https://tibble.tidyverse.org/articles/invariants.html#column-subsetting
+#' @export
+#' @noRd
+`[.polars_data_frame` <- function(x, i, j, drop = FALSE) {
+  cols <- names(x)
+
+  # useful for error messages below
+  i_arg <- substitute(i)
+  j_arg <- substitute(j)
+
+  # taken from tibble:::`[`
+  if (missing(i)) {
+    i <- NULL
+    i_arg <- NULL
+  } else if (is_null(i)) {
+    i <- integer()
+  }
+  if (missing(j)) {
+    j <- NULL
+    j_arg <- NULL
+  } else if (is_null(j)) {
+    j <- integer()
+  }
+  n_real_args <- nargs() - !missing(drop)
+  if (n_real_args <= 2L) {
+    if (!missing(drop)) {
+      warn(
+        c(
+          "!" = "`drop` argument ignored for subsetting a frame with `x[j]`.",
+          "i" = "It has an effect only for `x[i, j]`."
+        )
+      )
+      drop <- FALSE
+    }
+    j <- i
+    i <- NULL
+    j_arg <- i_arg
+    i_arg <- NULL
+  }
+
+  if (is_null(i) && is_null(j)) {
+    return(x)
+  }
+
+  #### Rows -----------------------------------------------------
+
+  # check accepted types for subsetting rows
+  if (!is_null(i) && !is_character(i) && !is.numeric(i) && !is_logical(i)) {
```

Comment:
After a little more investigation, it may be appropriate to use rlang's `is_bare_` functions here. (i.e., because S3 class vectors should not be used here)